va_explorer/users/utils/utils.py

```diff
@@ -0,0 +1,397 @@
+    user_data = User.objects\
+    .select_related('location_restrictions')\
+    .select_related('groups')\
+    .select_related('user_permissions')\
+    .values(
+            'uuid',
+            'is_active',
+            'is_staff',
+            role = F('groups__name'),
+            locations = F('location_restrictions__name'), 
+            permissions = F('user_permissions__codename')         
+    )
+    user_df = pd.DataFrame.from_records(user_data).rename(columns={'locations': 'location_restrictions'})
+    user_perms = (user_df[['uuid', 'permissions']]
+        .pivot_table(index='uuid', columns='permissions', values='uuid', aggfunc=lambda x: len(x) > 0)
+        .fillna(False)
+        .reset_index()
+        )
+    user_df = user_df.drop(columns="permissions").drop_duplicates().merge(user_perms, how="left")
+
+# update field worker usernames by matching names against all known interviewer names from VAs
+def link_fieldworkers_to_vas(emails=None, debug=False, match_threshold=80):
+    user_objects = User.objects
+    if emails:
+        emails = [emails] if type(emails) is str else emails
+        user_objects = user_objects.filter(email__in=emails)
+    # get list of field worker user names (from Users)
+    field_workers = filter(lambda x: x.is_fieldworker() and not x.name.startswith('Demo'), user_objects.all())
+    name_to_user = {user.name.lower().replace(' ', '_'): user for user in list(field_workers)}
+
+    # keep a list of va worker name keys for matching below
+    va_worker_keys = get_va_worker_names()
+    va_worker_names = list(va_worker_keys.keys())
+
+    # convert va names to lowercase and replace underscores with spaces
+    # fuzzy-match unique user_worker_names against unique va_worker_names
+    user_names = list(name_to_user.keys())
+    matches = [(user_name, fuzzy_match(user_name.lower(), va_worker_names, threshold=match_threshold)) for user_name in user_names]
+    # filter out tags that don't match any user names
+    matches = list(filter(lambda x: x[1], matches))
+    updated_va_ct = 0
+    if len(matches) > 0:
+        # update usernames of matching users
+        name_dict = {user_name: va_user_name for user_name, va_user_name in matches}
+        for name, new_username in name_dict.items():
+            # update Username
+            user = name_to_user[name]
+            user.username = new_username
+            user.set_va_username(new_username)
+            user.save()
+
+            if debug:
+                print(f"updating user {name}'s username to {new_username}")
+
+            # Make sure all VAs with matching field worker name (Id10010) are tagged with new username
+            # get original field worker tag for query (before it was lower-cased)
+            va_worker_key = va_worker_keys[new_username]
+            worker_vas = VerbalAutopsy.objects.filter(Id10010=va_worker_key)
+            assign_va_usernames(worker_vas, [new_username], override=True)
+
+            # for va in worker_vas:
```

Comment:
I'm thinking this is just a moved file, but is this comment still needed?

Reply:
oh, whoops! I meant to delete this file. I basically split the user `utils` into three separate files to prevent dependency issues. So I'll get rid of the file